File: backend/CMFragrances.API/Models/Perfume.cs
namespace CMFragrances.API.Models
{
    public class Perfume
    {
        public int Id { get; set; }

        public string Nombre { get; set; } = string.Empty;

        public string Marca { get; set; } = string.Empty;

        public string? Descripcion { get; set; }

        public string? Concentracion { get; set; }

        public int ContenidoML { get; set; }

        public decimal Precio { get; set; }

        public int Stock { get; set; }

        public string? Imagen { get; set; }

        public bool Activo { get; set; }

        public int CategoriaId { get; set; }

        public Categoria Categoria { get; set; } = null!;

        public ICollection<DetalleCarrito> DetalleCarritos { get; set; } = new List<DetalleCarrito>();

        public ICollection<DetallePedido> DetallePedidos { get; set; } = new List<DetallePedido>();
    }
}

File: frontend/cmfragrances-web/src/app/core/services/perfume.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';


export interface Perfume {

    id: number;

    nombre: string;

    marca: string;

    descripcion: string;

    concentracion: string;

    contenidoML: number;

    precio: number;

    stock: number;

    imagen: string;

    activo: boolean;

    categoriaId: number;

    categoria: string;

}


export interface PerfumeRequest {

    nombre: string;

    marca: string;

    descripcion: string;

    concentracion: string;

    contenidoML: number;

    precio: number;

    stock: number;

    imagen: string;

    activo: boolean;

    categoriaId: number;

}


@Injectable({
    providedIn: 'root'
})
export class PerfumeService {


    // ==========================================
    // HTTP
    // ==========================================

    private http =
        inject(HttpClient);


    // ==========================================
    // URL DE LA API
    // ==========================================

    private apiUrl =
        environment.apiUrl;


    // ==========================================
    // OBTENER TODOS LOS PERFUMES
    // ==========================================

    obtenerPerfumes(): Observable<Perfume[]> {

        return this.http.get<Perfume[]>(

            `${this.apiUrl}/Perfumes`

        );

    }


    // ==========================================
    // OBTENER PERFUME POR ID
    // ==========================================

    obtenerPerfume(
        id: number
    ): Observable<Perfume> {

        return this.http.get<Perfume>(

            `${this.apiUrl}/Perfumes/${id}`

        );

    }


    // ==========================================
    // CREAR PERFUME
    // ==========================================

    crearPerfume(
        perfume: PerfumeRequest
    ): Observable<Perfume> {

        return this.http.post<Perfume>(

            `${this.apiUrl}/Perfumes`,

            perfume

        );

    }


    // ==========================================
    // ACTUALIZAR PERFUME
    // ==========================================

    actualizarPerfume(
        id: number,
        perfume: PerfumeRequest
    ): Observable<void> {

        return this.http.put<void>(

            `${this.apiUrl}/Perfumes/${id}`,

            perfume

        );

    }


    // ==========================================
    // ELIMINAR PERFUME
    // ==========================================

    eliminarPerfume(
        id: number
    ): Observable<void> {

        return this.http.delete<void>(

            `${this.apiUrl}/Perfumes/${id}`

        );

    }

}